### DISCUSSION TITLE
Manually setting corner radius of menu's blurred area in kvconfig?

### DISCUSSION THREAD

**Discussion by doncsugar** [tsujan/Kvantum#761] [Ideas]
It appears that we have reached critical mass and are willing to try anything to resolve "Korners" over in KDE. As seen [here](https://invent.kde.org/plasma/kwin/-/merge_requests/1961), an early suggestion would be to allow manually setting the radius of the blurred area behind Aurorae decorations. Whilst hacky and will fail to support more extravagant styles, I believe the benefits to the many (and more commonly used) simple rounded styles will benefit from a clean appearance—as will users.
![image](https://user-images.githubusercontent.com/46570709/152670087-28595504-46d8-4f38-81d7-9aaf28e9d34c.png)
For themes that have very high values for the radius, this unsettling contrast between blur and menu is even more noticeable. Is it possible for such functionality to be implemented into Kvantum?

**Comment by tsujan** [comment]
> Is it possible for such functionality to be implemented into Kvantum?

Kvantum doesn't depend on `kdecoration`. So, those methods aren't related to it.

The only rounded corners in top level windows that are controlled by Kvantum are in menus. The rest is controlled by KWin (if present). Unlike buttons and some other elements, a menu SVG element shouldn't have very rounded corners in Kvantum.

It may be possible to add a key like `menu_blur_roundness` and choose its value by trial and error for a theme that has rounded menus. But it might need lots of code lines for such a small effect (and only under KWin) and I'm not sure it'll be CPU-friendly. If I find the time, I might do some experiments.

**Reply by doncsugar** [reply]
> If I find the time, I might do some experiments.

Not to dissuade you from experimenting, but there may be a medium solution that could reduce the effect without requiring any changes to themes. I once recall the way the blur area was calculated being switched to only blur the pixels immediately under it. Like in the picture above, the corners of the blurred region are white because the content outside of it that is being blurred is white. If it is possible, restricting the considered area to the menu area or a smaller region may reduce the unsightly phenomenon. My concern with restricting the considered area to just the menu itself is that it may be jarring if content under it moves, as having a single pixel of white may suddenly cause a large area of the menu to become white.

Ultimately, this suggestion is flawed if the contrast effect is in use, which can alter the appearance of the blurred region, regardless.

> Unlike buttons and some other elements, a menu SVG element shouldn't have very rounded corners in Kvantum.

I agree, though that may not be what happens in practice. I believe radii go up to 12px, though 8-10 is likely a more appropriate max. Since the padding on the top and bottom increases with the radius of the menu corners and makes menus look quite strange, supporting a lower max makes sense.

**Reply by tsujan** [reply]
In Kvantum, the whole SVG element of a translucent menu can have quite rounded corners (as it does in most themes included in Kvantum) but most of that roundness belongs to the menu shadow, which is excluded from blurring. Moreover, "corner blurring" depends on the contrast that may exist in the background itself and isn't always visible. Usually just a few pixels at the corners may need to be excluded. This screenshot shows the "worst" thing that may happen (at the top corners):

![menu_corners](https://user-images.githubusercontent.com/981076/152746135-5b83054d-127b-41e4-a292-4d552c8b4a68.png)

However, the exact number of pixels can't be calculated because it depends on the SVG elements.

**Reply by doncsugar** [reply]
> However, the exact number of pixels can't be calculated because it depends on the SVG elements.

What methods are there to determine the number of pixels? Is the inverse of "full" pixels for a given radius enough? 
![image](https://user-images.githubusercontent.com/46570709/153141803-24d6293a-9da1-45bd-8354-0e5fd020f90f.png)
Are there other manual ways of handling it than supplying a radius for Kvantum to implement?

**Reply by tsujan** [reply]
The number of pixels should be derived from the radius. That's math and isn't a problem (although it may take time to find the best code). The probable problem is about the extra CPU usage caused by changing a single blurred rectangle to a region consisting of several rectangles — in your screenshot, 21.

**Reply by doncsugar** [reply]
Which screenshot has 21? 21 rectangles? 21 pixels to be omitted?

Also, I'm not keen on how the code works, but could you say which of these three pictures illustrates what you mean by "several rectangles"?
![image](https://user-images.githubusercontent.com/46570709/153296424-3eacdf73-b22b-4160-9bd9-201d3cca0684.png)
I cannot figure out where the 21 comes from.

**Reply by tsujan** [reply]
None. I was referring to the whole "rounded" rectangle, not just its top left corner.

Anyway, it's implemented now.

**Comment by tsujan** [comment]
OK, I tested and it worked fine. This is without "blur rounding" and with a scale factor of 2:

![before](https://user-images.githubusercontent.com/981076/153214656-f3f6e949-bdb1-4329-9e1d-fdf13ed9d041.png)

And this with a "blur radius" of 2:

![after](https://user-images.githubusercontent.com/981076/153214717-ca504f9c-aafe-400c-b7b4-e68a0386a1c7.png)

I saw no tangible difference in CPU usage, but I have a powerful laptop. However, it'll be optional.

EDIT: The blur radius is only 2 but because the scale factor is 2, it seems to be 4.

**Reply by doncsugar** [reply]
The entire region seems to be shifted down 1 pixel in both of these images.
![image](https://user-images.githubusercontent.com/46570709/153297144-ed7e1504-7e56-4516-b629-8a6d3a69c8ef.png)

I will do a rudimentary benchmark on my notoriously slow machine for a few radii and let you know if I see any performance issues related to high radii there. For reference, what are your specifications?

I will also check for radius values that will best hide the sharp corners:
![image](https://user-images.githubusercontent.com/46570709/153297981-b0df1e9e-13ee-4adf-a3c4-0764d76d436c.png)

One last concern regarding an issue I spoke with you about: Now that a radius for the blur is specified, does that mean when a user reduces the opacity of the menu, the discrepancy between the shadow and the shadow inside the menu's rectangle will be more obvious?

**Reply by tsujan** [reply]
> The entire region seems to be shifted down 1 pixel in both of these images.

No. It's related to anti-aliasing.

**Reply by tsujan** [reply]
> Now that a radius for the blur is specified, does that mean when a user reduces the opacity of the menu, the discrepancy between the shadow and the shadow inside the menu's rectangle will be more obvious?

If you mean 4 tiny parts of the shadow outside rounded corners, no. For now, I think it isn't worth adding code lines.

**Comment by tsujan** [comment]
Done in https://github.com/tsujan/Kvantum/commit/d87dcf284081876743b58abc379ec32e22fce752

Usually, a value of 2 is enough for menus/tooltips that have rounded corners.

Thanks for the discussion!

**Reply by doncsugar** [reply]
I have tested it. I like that you incorporated the radius into the Manager.

Why do you suppose the other corners are being squashed? The screenshot is at radius 10:
![image](https://user-images.githubusercontent.com/46570709/153352533-4d5fa1f5-4d1c-493c-8774-bb0f0cfdcd2f.png)

Tooltip _may_ exhibit odd behavior if the radius is larger than half its height.

**Reply by tsujan** [reply]
> Tooltip may exhibit odd behavior if the radius is larger than half its height.

The code doesn't allow a radius > 10.

EDIT: → https://github.com/tsujan/Kvantum/commit/8e79ca8f11fe8e73c2ec3c51b4919cccb2580c06
